nomad/node_endpoint.go

```diff
@@ -1062,39 +1062,54 @@ func (n *Node) UpdateAlloc(args *structs.AllocUpdateRequest, reply *structs.Gene
 	now := time.Now()
 	var evals []*structs.Evaluation
 
-	for _, alloc := range args.Alloc {
-		alloc.ModifyTime = now.UTC().UnixNano()
+	for _, allocToUpdate := range args.Alloc {
+		allocToUpdate.ModifyTime = now.UTC().UnixNano()
+
+		if !allocToUpdate.TerminalStatus() {
+			continue
+		}
+
+		alloc, _ := n.srv.State().AllocByID(nil, allocToUpdate.ID)
+		if alloc == nil {
+			continue
+		}
+
+		job, err := n.srv.State().JobByID(nil, alloc.Namespace, alloc.JobID)
+		if err != nil {
+			n.logger.Error("UpdateAlloc unable to find job", "job", alloc.JobID, "error", err)
+			continue
+		}
+		if job == nil {
```

Comment:
Note to reviewers: I'm not sure I totally understand the circumstances that could lead this this being true or why we log that differently rather than doing as `if err != nil || job == nil`, or why a `nil` alloc can pass without a log entry but a `nil` job can't.

But this section is a few years old and I'd rather not change it in this PR without a better notion of why we do it that way.

Reply:
It's interesting, because it seems like there's a race here. If an alloc is running at the time a job is updated, we could be fetching either the new job and applying logic from the updated taskgroup to this alloc or nothing if the job was deleted. I would have expected that we'd fetch the job using the version here.